evm/client/rpc_client.go

```diff
@@ -180,48 +162,33 @@ func (r *RPCClient) Ping(ctx context.Context) error {
 	return err
 }
 
-func (r *RPCClient) UnsubscribeAllExcept(subs ...multinode.Subscription) {
-	r.subsSliceMu.Lock()
-	defer r.subsSliceMu.Unlock()
-
-	keepSubs := map[multinode.Subscription]struct{}{}
-	for _, sub := range subs {
-		keepSubs[sub] = struct{}{}
-	}
-
-	for sub := range r.subs {
-		if _, keep := keepSubs[sub]; !keep {
-			sub.Unsubscribe()
-			delete(r.subs, sub)
-		}
-	}
-}
-
 // Not thread-safe, pure dial.
 func (r *RPCClient) Dial(callerCtx context.Context) error {
-	ctx, cancel := r.makeQueryCtx(callerCtx, r.rpcTimeout)
+	ctx, cancel, _ := r.AcquireQueryCtx(callerCtx, r.rpcTimeout)
 	defer cancel()
 
-	if r.ws == nil && r.http == nil {
+	if r.ws.Load() == nil && r.http.Load() == nil {
 		return errors.New("cannot dial rpc client when both ws and http info are missing")
 	}
 
 	promEVMPoolRPCNodeDials.WithLabelValues(r.chainID.String(), r.name).Inc()
 	lggr := r.rpcLog
-	if r.ws != nil {
-		lggr = lggr.With("wsuri", r.ws.uri.Redacted())
-		wsrpc, err := rpc.DialWebsocket(ctx, r.ws.uri.String(), "")
+	ws := r.ws.Load()
+	if ws != nil {
+		lggr = lggr.With("wsuri", ws.uri.Redacted())
+		wsrpc, err := rpc.DialWebsocket(ctx, ws.uri.String(), "")
 		if err != nil {
 			promEVMPoolRPCNodeDialsFailed.WithLabelValues(r.chainID.String(), r.name).Inc()
-			return r.wrapRPCClientError(pkgerrors.Wrapf(err, "error while dialing websocket: %v", r.ws.uri.Redacted()))
+			return r.wrapRPCClientError(pkgerrors.Wrapf(err, "error while dialing websocket: %v", ws.uri.Redacted()))
 		}
 
-		r.ws.rpc = wsrpc
-		r.ws.geth = ethclient.NewClient(wsrpc)
+		ws.rpc = wsrpc
```

Comment:
ws.rpc and ws.geth can be accessed concurrently. Instead of modifying ws structure we should store a new one. 
```
r.ws.Store(&rawclient{uri: ws.uri, rpc: wsrpc, get: ethclient.NewClient(wsrpc)})
```